src/dma.md

```diff
@@ -0,0 +1,348 @@
+involves `buf` so such reorderings do *not* result in undefined behavior.
+
+Note that `compiler_fence` is a bit stronger than what's required. For example,
+the fences will prevent the operations on `x` from being merged even though we
+know that `buf` doesn't overlap with `x` (due to Rust aliasing rules). However,
+there exist no intrinsic that's more fine grained than `compiler_fence`.
+
+> **NOTE:** You *may* need to use a [`atomic::fence`][] (a memory barrier)
+> instead of a `compiler_fence` if you expect your API to be used with buffers
+> shared between different cores. But this depends on how you design your API.
+
+[`atomic::fence`]: https://doc.rust-lang.org/core/sync/atomic/fn.fence.html
+
+## Generic buffer
+
+Our API is more restrictive that it needs to be. For example, the following
+program won't be accepted even though it's valid.
+
+``` rust
+{{#include ../ci/dma/examples/five.rs:67:85}}
+```
+
+To accept such program we can make the buffer argument generic.
+
+``` rust
+{{#include ../ci/dma/examples/five.rs:9:65}}
+```
+
+> **NOTE:** `AsRef<[u8]>` (`AsMut<[u8]>`) could have been used instead of
+> `AsSlice<Element = u8>` (`AsMutSlice<Element = u8`).
+
+Now the `reuse` program will be accepted.
+
+## Immovable buffers
+
+With this modification the API will also accept arrays by value (e.g. `[u8;
+16]`). However, using arrays can result in pointer invalidation. Consider the
+following program.
+
+``` rust
+{{#include ../ci/dma/examples/five.rs:88:103}}
+{{#include ../ci/dma/examples/five.rs:105:112}}
+```
+
+The `read_exact` operation will use the address of the `buffer` local to the
+`start` function. That local `buffer` will be freed when `start` returns and the
+pointer used in `read_exact` will become invalidated. You'll end up with a
+situation similar to the [`unsound`](#dealing-with-memforget) example.
+
+To avoid this problem we require that the buffer used with our API retains its
+memory location even when it's moved. The [`Pin`] newtype provides such
+guarantee. We can update our API to required that all buffers are "pinned"
+first.
+
+[`Pin`]: https://doc.rust-lang.org/nightly/std/pin/index.html
+
+> **NOTE:** To compile all the programs below this point you'll need Rust
+> `>=1.33.0`. As of time of writing (2019-01-04) that means using the nightly
+> channel.
+
+``` rust
+{{#include ../ci/dma/examples/six.rs:16:33}}
+{{#include ../ci/dma/examples/six.rs:48:59}}
+{{#include ../ci/dma/examples/six.rs:74:75}}
+```
+
+> **NOTE:** We could have used the [`StableDeref`] trait instead of the `Pin`
+> newtype but opted for `Pin` since it's provided in the standard library.
+
+[`StableDeref`]: https://crates.io/crates/stable_deref_trait
+
+With this new API we can use `&'static mut` references, `Box`-ed slices, `Rc`-ed
+slices, etc.
+
+``` rust
+{{#include ../ci/dma/examples/six.rs:78:89}}
+{{#include ../ci/dma/examples/six.rs:91:101}}
+```
+
+## `'static` bound
```

Comment:
If we have a 'static lifetime, do we still need to Pin the buffers? It seems like adding 'static lifetime solves the original problem of using stack-based buffers, which don't live long enough.

If you can provide an example of an issue that would occur when you only have 'static without Pin, I think that would help me understand better.

Thank you!

Reply:
We recently discovered that not even `'static` + `Pin` is enough, see https://github.com/rust-embedded/embedonomicon/issues/64, we are working on a new documentation for this, see https://github.com/ra-kete/dma-poc for more details.

Reply:
Thank you for the response. The links where very helpful, and I think the new documentation is very clear